Simon-Initiative/oli-torus
# [FEATURE] Enforce paywalled sections, allow payment code redemption and admin management CMU-191

Represents P5 of work from Paywall FDD.

Closes #1555

This work adds in the enforcement of paywalled sections, the ability for a user to redeem a payment code to bypass that the paywall, and an admin only feature to view the issued payment codes and to create and download new batches of codes. 

There is some infra work here to refine the reusable `SortableTableModel` and the macro that allows reuse of the common table operation handlers.  I refined the extensibility approach to use module attrs.  I also had to redo the custom sort approach to make it rely on the more flexible `Enum.sort_by/3`. 

To test this out, do the following:

1. Create a course PRODUCT and require payment and allow a grace period.
2. Create an open and free course section (or launch from LMS and create regular one) and base it on the paid product. Note, to use grace period with "relative to section" strategy you must select a start date for this section. 
3. As admin, go to this product and generate some payment codes.
4. Then, in new browser as student attempt to access the O+F course.  You should be blocked and redirected to the page prompting you to either make a payment or supply a payment code.
5. Supply one of the payment codes and, voilà, you now have access.
6. Verify the payment code shows up as having been applied in the admin payments interface.
